README


This archive contains the distribution Design-Creation,
version 0.047:

  Create Knock-Out Designs

This software is copyright (c) 2013 by Genome Research Ltd.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
{{$NEXT}}

0.047     2018-05-18 14:31:36 Europe/London

    Allow num_mismatches to be zero

0.046     2018-02-13 09:29:28 Europe/London

0.045     2017-02-22 15:21:44 Europe/London
    Wrapper scratch migration
0.044     2017-02-21 17:01:50 Europe/London
    Switch from scratch109 to scratch117
0.043     2017-02-06 11:22:28 Europe/London

    Switch from scratch109 to scratch117

0.042     2016-09-12 10:38:03 Europe/London
    Fusion deletion tweaks
0.041     2016-02-05 14:25:26 Europe/London
    Fixed bugs with the fusion-deletion generation
0.040     2016-01-04 16:14:58 Europe/London

0.039     2015-12-16 12:23:35 Europe/London
    Added fusion-deletion design type

0.038     2015-12-07 16:43:12 Europe/London

0.037     2015-05-20 07:45:34 Europe/London

    Update scripts and Constants file.

0.036     2015-05-14 11:49:30 Europe/London

    Move farm wrapper script to bin dir.
    Remove old document files.
    Add reference to new github wiki.
    Clean up some POD.

0.035     2015-03-17 13:57:40 Europe/London

    Primer3 Util: Add primer_task, sequence_excluded_region and sequence_included_region options

0.034     2015-02-24 13:06:57 Europe/London

    Switch to using WebAppCommon::Util::EnsEMBL

0.033     2015-01-19 10:42:57 Europe/London

    Scratch109 back up, use this 

0.032     2014-12-12 09:04:42 Europe/London

    Scratch109 offline for upgrade, use different file system

0.031     2014-12-08 09:09:17 Europe/London

    Minor bugfixes.
    Add optional design comment

0.030     2014-10-07 14:31:52 Europe/London

    Update design code to work with new human assembly: GRCh38

0.029     2014-10-02 09:03:59 Europe/London

    Bugfix: throw meaningful error when unexpected chromosome used

0.027     2014-07-11 13:57:59 Europe/London

    Various bugfixes and script improvements.
    Design ids data in design attempts table is not a postgres array, so send in array ref.

0.026     2014-05-22 08:42:37 Europe/London

    Add option to allow oligos with n off target hits to be validated.

0.025     2014-04-24 15:16:26 Europe/London

    Re-release, did not merge changes for 0.024 release

0.024     2014-04-17 13:47:27 Europe/London

    Add extra options to Exonerate wrapper

0.023     2014-03-19 07:39:37 Europe/London

    Refactor of command structure, now split into step and complete commands.
    Update design attempts with candidate oligos and candidate regions.
        - this gives the users some oligos even if the design fails

0.022     2014-03-11 13:20:10 Europe/London

    Store all command parameters when creating initial design attempt record.
    Throw custom error class when we do not have any valid primer pairs for a region.
    Change default offsets in gibson deletion designs.

0.021     2014-02-19 08:21:47 Europe/London

    Write new Primer3 config values into parameters file.
    Throw a more meaningful error when Primer3 dies.
    Can set Primer3 gc_clamp value.

0.020     2014-02-17 14:19:24 Europe/London

    Added option to set Primer3 config values by command line for: 
        - melting temp
        - gc percentage
        - primer size

0.019     2014-02-13 08:48:42 Europe/London

    Refactor to add a initial target step.
    Create location specified gibson design commands.

0.018     2014-02-03 09:12:24 Europe/London
    
    Add deletion gibson design creation

0.017     2014-01-29 14:51:18 Europe/London

    Allow multi exon targeting for gibson designs

0.016     2014-01-21 07:45:20 Europe/London

    Switch ref genomes for bwa and blast to scratch109

0.015     2013-12-17 10:32:57 Europe/London

    Bugfix: check for existance of optional design parameter

0.014     2013-12-17 09:14:56 Europe/London

    Move create-multiple-designs.pl script back into /bin

0.013     2013-12-12 10:25:26 Europe/London

    Moved scripts from /bin to /script

0.012     2013-12-12 09:54:54 Europe/London

    Throw better errors with more specific DesignCreate::Exception classes.
    Create a fail.yaml file with detailed reasons a design failed.
    Create and update design_attempt records if persisting design.
    Clean up design directory before starting.
    Add option to specify number of bwa threads to use.

0.011     2013-09-25 13:58:50 Europe/London

    Add util module to wrap around bwa and samtools.
    Use bwa instead of exonerate when checking genomic specificity of gibson oligos.
    Add more checks for gibson oligos: not too close to other exons.

0.010     2013-08-27 11:05:19 Europe/London

    Gibson design creation
    Design target scripts
    Code refactor
    Store design parameters in LIMS2
    Multiple bugfixes and todo's done
    Updated POD docs

0.009     2013-05-23 07:11:28 Europe/London

    Set EnsEMBL registry to try to reconnect when connection lost.
    When resetting registry first disconnect all connections to the the database.

0.008     2013-05-22 10:56:57 Europe/London

    Bugfix for resetting ensembl connection.
    Reset the ensembl registry properly, it is a class attribute.

0.007     2013-05-22 08:13:30 Europe/London

    Added more try / catch blocks to isolate the EnsEMBL API disconnect error.
    Minor changes to reporting and run multiple designs scripts.

0.006     2013-05-21 07:48:30 Europe/London

    Added script for finding human design targets.
    Add option to specify classes of repeat masking to apply to sequence.
    Re-try connecting to EnsEMBL API if there are any errors ( up to 5 times )
    Various bug fixes and refactoring.

0.005     2013-05-07 12:31:37 Europe/London

    Add species option, can now design on Mouse and Human targets.
    Add del-exon-design command, creates design based on target exon.

0.004     2013-04-23 07:50:16 Europe/London

    For conditional designs add optional block oligo region overlap value.


lib/DesignCreate/Util/BWA.pm
package DesignCreate::Util::BWA;

=head1 NAME

DesignCreate::Util::BWA

=head1 DESCRIPTION

Align sequence(s) against a genome to find number of hits using BWA aln

=cut

use Moose;
use DesignCreate::Exception;
use DesignCreate::Exception::MissingFile;
use DesignCreate::Types qw( PositiveInt NaturalNumber Species );
use DesignCreate::Constants qw(
    $BWA_CMD
    $SAMTOOLS_CMD
    $XA2MULTI_CMD
    %BWA_GENOME_FILES
);
use WebAppCommon::Util::RemoteFileAccess;
use MooseX::Types::Path::Class::MoreCoercions qw/AbsFile/;
use Path::Class  qw( file );
use YAML::Any qw( LoadFile DumpFile );
use IPC::Run 'run';
use Const::Fast;
use namespace::autoclean;
use Bio::SeqIO;
use IO::String;
use Data::Dumper;
use Path::Class;
use Data::UUID;

with qw( MooseX::Log::Log4perl );

has primers => (
    is => 'ro',
    isa => 'HashRef',
);

has api => (
    is => 'ro',
    isa => 'WebAppCommon::Util::RemoteFileAccess',
    lazy_build => 1,
);

sub _build_api {
    return WebAppCommon::Util::RemoteFileAccess->new({server => 'bwa'});
}

has work_dir => (
    is         => 'ro',
    isa        => 'Path::Class::Dir',
    lazy_build => 1,
);

sub _build_work_dir {
    my $self = shift;
    my $unique_string = Data::UUID->new()->create_str();
    my $root_dir = $ENV{'LIMS2_BWA_OLIGO_DIR'} // '/var/tmp/bwa';
    my $work_dir = dir($root_dir, '_' . $unique_string);
    $self->api->make_dir($work_dir->stringify);
    #or croak 'Could not create directory ' . $self->work_dir->stringify . ": $!";
    return $work_dir;
}

has query_file => (
    is         => 'ro',
    isa        => 'Path::Class::File',
    lazy_build => 1,
);

sub _build_query_file {
    my $self = shift;
    my $file_content = '';
    my $file_content_io = IO::String->new($file_content);
    my $seq_out = Bio::SeqIO->new( -fh => $file_content_io, -format => 'fasta' );
    if ($self->primers->{'left'} || $self->primers->{'right'}) {
        foreach my $oligo ( sort keys %{$self->primers->{'left'}} ) {
            my $fasta_seq = Bio::Seq->new(
                -seq => $self->primers->{'left'}->{$oligo}->{seq},
                -id  => $oligo
            );
            $seq_out->write_seq($fasta_seq);
        }
        foreach my $oligo ( sort keys %{$self->primers->{'right'}} ) {
            my $fasta_seq = Bio::Seq->new(
                -seq => $self->primers->{'right'}->{$oligo}->{seq},
                -id  => $oligo
            );
            $seq_out->write_seq($fasta_seq);
        }
    } else {
        foreach my $oligo ( sort keys %{$self->primers} ) {
            my $fasta_seq = Bio::Seq->new(
                -seq => $self->primers->{$oligo}->{seq},
                -id  => $oligo
            );
            $seq_out->write_seq($fasta_seq);
        }
    }
    my $fasta_file_name = $self->work_dir->file('oligos.fasta');
    $self->api->post_file_content($fasta_file_name, $file_content);
    return $fasta_file_name;
}

has species => (
    is       => 'ro',
    isa      => Species,
    required => 1,
);

has num_bwa_threads => (
    is     => 'ro',
    isa     => PositiveInt,
    default => 2,
);

# default of 2 only gets hits with > 90% similarity
has num_mismatches => (
    is       => 'ro',
    isa      => NaturalNumber,
    required => 1,
    default  => 2,
);

has three_prime_check => (
    is      => 'ro',
    isa     => 'Bool',
    default => 0,
);

has delete_bwa_files => (
    is      => 'ro',
    isa     => 'Bool',
    default => 1,
);

has target_file => (
    is         => 'rw',
    isa        => 'Path::Class::File',
    lazy_build => 1,
);

sub _build_target_file {
    my $self = shift;

    my $file = file( $BWA_GENOME_FILES{ $self->species } );
    return $file->absolute;
}

has sam_multi_file => (
    is         => 'ro',
    isa        => AbsFile,
    lazy_build => 1,
);

sub _build_sam_multi_file {
    return shift->work_dir->file('query.multi.sam')->absolute;
}

has bed_file => (
    is         => 'ro',
    isa        => AbsFile,
    lazy_build => 1,
);

sub _build_bed_file {
    return shift->work_dir->file('query.bed')->absolute;
}

# oligo seqs, all on +ve strand
has oligo_seqs => (
    is                => 'ro',
    isa               => 'HashRef',
    lazy_build        => 1,
    traits            => [ 'Hash' ],
    handles           => {
        get_oligo_seq => 'get',
    }
);

sub _build_oligo_seqs {
    my $self = shift;
    my %oligo_seqs;

    #my $file_content = $self->api->get_file_content($self->query_file);
    #print "$file_content\n";
    #my $file_content_io = IO::String->new($file_content);
    #my $seq_in = Bio::SeqIO->new( -fh => $file_content_io, -format => 'fasta' );
    my $seq_in = Bio::SeqIO->new( -fh => $self->query_file->openr, -format => 'fasta' );
    while ( my $seq = $seq_in->next_seq ) {
        $oligo_seqs{ $seq->display_id } = $seq->seq;
    }

    return \%oligo_seqs;
}

has matches => (
    is  => 'rw',
    isa => 'HashRef',
);

=head2 run_bwa_checks

Run bwa alignments along with multiple other steps to parse this output.
Generate number of hits a oligo has against the genome.

=cut
sub run_bwa_checks {
    my $self = shift;
    $self->log->info( 'Running bwa alignment checks' );

    $self->generate_sam_file;
    my $oligo_hits = $self->oligo_hits;
    #NOTE This is not implemented
    if ( $self->three_prime_check ) {
        $self->generate_bed_file;
        $self->oligo_hits_three_prime_check( $oligo_hits );

        my $three_prime_check_file = $self->work_dir->file( 'three_prime_check.yaml' );
        DumpFile( $three_prime_check_file, $self->matches );
    }
    else {
        # the basic oligo hits info is good enough
        $self->matches( $oligo_hits );
        $self->sam_multi_file->remove if $self->delete_bwa_files;
    }

    $self->bed_file->remove if $self->delete_bwa_files;

    return;
}

=head2 generate_sam_file

Run the aln and samse steps of bwa, output is a sam file.
* bwa aln - perform alignments of oligo sequences against whole genome
* bwa samse - convert sai file from bwa aln step into a sam file
* xa2multi - put each hit for oligo into seperate line in sam file

=cut
sub generate_sam_file {
    my $self = shift;
    $self->log->info( 'Generating sam file' );

    my @aln_command = (
        $BWA_CMD,
        'aln',                         # align command
        "-n", $self->num_mismatches,   # number of mismatches allowed over sequence
        "-o", 0,                       # disable gapped alignments
        "-N",                          # disable iterative search so we get all hits
        "-t", $self->num_bwa_threads,  # specify number of threads
        $self->target_file->stringify, # target genome file, indexed for bwa
        $self->query_file->stringify,  # query file with oligo sequences
    );
    $self->log->debug( "BWA aln command: " . join( ' ', @aln_command ) );

    my $bwa_aln_file = $self->work_dir->file('query.sai')->absolute;
    my $bwa_aln_log_file = $self->work_dir->file( 'bwa_aln.log' )->absolute;
    my $target_file = $self->target_file->stringify;
    my $query_file = $self->query_file->stringify;
    system("ssh bwa \"bwa aln -n 0 -o 0 -N -t 2 $target_file $query_file > $bwa_aln_file 2> $bwa_aln_log_file\"");
    #run( \@aln_command,
    #     '<', \undef,
    #    '>', $bwa_aln_file->stringify,
    #    '2>', $bwa_aln_log_file->stringify
    #) or DesignCreate::Exception->throw(
    #        "Failed to run bwa aln command, see log file: $bwa_aln_log_file" );

    my @samse_command = (
        $BWA_CMD,
        'samse',                       # converts sai file (binary) to sam file
        '-n 900000',                   # max number of allowed hits per oligo
        $self->target_file->stringify, # target genome file
        $bwa_aln_file->stringify,      # sai binary file, output from bwa aln step
        $self->query_file->stringify,  # query file with oligo sequences
    );
    $self->log->debug( "BWA samse command: " . join( ' ', @samse_command ) );

    my $sam_file = $self->work_dir->file('query.sam')->absolute;
    my $bwa_samse_log_file = $self->work_dir->file( 'bwa_samse.log' )->absolute;
    system("ssh bwa \"bwa samse -n 900000 $target_file $bwa_aln_file $query_file > $sam_file 2> $bwa_samse_log_file\"");
    #run( \@samse_command,
    #    '<', \undef,
    #    '>', $sam_file->stringify,
    #    '2>', $bwa_samse_log_file->stringify
    #) or DesignCreate::Exception->throw(
    #        "Failed to run bwa samse command, see log file: $bwa_samse_log_file" );

    my @xa2multi_command = (
        $XA2MULTI_CMD,
        $sam_file->stringify, # sam file output from samse step
    );
    $self->log->debug( "xa2multi command: " . join( ' ', @xa2multi_command ) );

    my $xa2multi_log_file = $self->work_dir->file('xa2multi.log')->absolute;
    my $err = "";
    my $sam_multi_file = $self->sam_multi_file->stringify;
    system("ssh bwa \"/home/ubuntu/xa2multi.pl $sam_file > $sam_multi_file 2> $xa2multi_log_file\"");
    #run( \@xa2multi_command, '<', \undef, '>', $self->sam_multi_file->stringify, '2>', \$err)
    #    or DesignCreate::Exception->throw(
    #        "Failed to run xa2multi command: $err" );

    if ( $self->delete_bwa_files ) {
        $sam_file->remove;
        $bwa_aln_file->remove;
    }

    return;
}

=head2 oligo_hits

Calculate number of hits each oligo got against the genome.
This is done by parsing the data from the sam multi file.

=cut
sub oligo_hits {
    my ( $self ) = @_;
    $self->log->info( 'Calculating oligo hits' );
    my %oligo_hits;

    my $file_content = $self->api->get_file_content($self->sam_multi_file);
    my $fh = IO::String->new($file_content);
    #my $fh = $self->sam_multi_file->openr;
    while ( <$fh> ) {
        next if /^@/;
        my @data = split /\t/;
        my $id           = $data[0];
        my $flag         = $data[1];
        my $chromosome   = $data[2];
        my $chr_start    = $data[3];
        my $score        = $data[4];
        my $sub_opt_hits = $data[14];

        if ( $self->_primary_alignment( $flag ) ) {
            if ( $self->_segment_unmapped( $flag ) ) {
                $oligo_hits{$id}{unmapped} = 1;
                next;
            }
            $oligo_hits{$id}{hits}++;

            $oligo_hits{$id}{score} = $score;
            $oligo_hits{$id}{sub_opt_hits} = $sub_opt_hits;
            $oligo_hits{$id}{chr} = $chromosome;
            $oligo_hits{$id}{start} = $chr_start;

            my $threshold = $ENV{BWA_GENOMIC_THRESHOLD} || 30;
            # score ok above 30 look to be totally unique
            if ( $score > $threshold ) {
                $oligo_hits{$id}{unique_alignment} = 1;
            }
        }
        else {
            $oligo_hits{$id}{hits}++;
            push @{ $oligo_hits{$id}{hit_locations} }, { chr => $chromosome, start => $chr_start };
        }
    }

    my $oligo_hits_file = $self->work_dir->file( 'oligo_hits.yaml' );
    $self->api->post_file_content($oligo_hits_file, Dumper(\%oligo_hits));
    #DumpFile( $oligo_hits_file, \%oligo_hits );

    return \%oligo_hits;
}

=head2 _segment_unmapped

Return true if segment unmapped SAM flag is true.

=cut
sub _segment_unmapped {
    my ( $self, $flag ) = @_;
    if ($flag & 0x4){
        return 1;
    }
    return;
}

=head2 _primary_alignment

Return true if secondary alignment SAM flag is false.

=cut
sub _primary_alignment {
    my ( $self, $flag ) = @_;
    if ($flag & 0x100){
        return;
    }
    return 1;
}

=head2 generate_bed_file

Take sam file output from previous steps and generate bed file.
* samtools view - convert sam file to bam file
* samtools sort - sort bam file by left most coordinate, in chromosome order
* bamToBed - convert bam file to bed file

=cut
sub generate_bed_file {
    my ( $self ) = @_;
    my ( $out, $err ) = ( "","" );
    $self->log->info( 'Generating bed file' );

    my @view_command = (
        $SAMTOOLS_CMD,
        'view',                           # convert same file to bam file
        '-b',                             # output in bam format
        '-S',                             # input is sam file
        $self->sam_multi_file->stringify, # sam file
    );
    $self->log->debug( "samtools view command: " . join( ' ', @view_command ) );

    $err = "";
    my $bam_file = $self->work_dir->file('query.bam')->absolute;
    run( \@view_command, '<', \undef, '>', $bam_file->stringify, '2>', \$err)
        or DesignCreate::Exception->throw(
            "Failed to run samtools view command: $err" );

    my $temp_file = $self->work_dir->file('query.sorted')->absolute;
    my @sort_command = (
        $SAMTOOLS_CMD,
        'sort',                         # sort by left most coordinate
        $bam_file->stringify,           # input bam file
        $temp_file->stringify,          # output - sorted bam file
    );
    $self->log->debug( "samtools sort command: " . join( ' ', @sort_command ) );

    $err = "";
    run( \@sort_command, '<', \undef, '>', \$out, '2>', \$err)
        or DesignCreate::Exception->throw(
            "Failed to run samtools sort command: $err" );

    my $sorted_bam_file = $self->work_dir->file('query.sorted.bam')->absolute;
    DesignCreate::Exception::MissingFile->throw( file => $sorted_bam_file, dir => $self->work_dir )
        unless $self->work_dir->contains( $sorted_bam_file );

    my @bamToBed_command = (
        'bamToBed',                        # convert bam to bed file
        "-i", $sorted_bam_file->stringify, # input bam file
    );
    $self->log->debug( "bamToBed command: " . join( ' ', @bamToBed_command ) );

    run( \@bamToBed_command, '<', \undef, '>', $self->bed_file->stringify, '2>', \$err,)
        or DesignCreate::Exception->throw(
            "Failed to run bamToBed command: $err" );

    if ( $self->delete_bwa_files ) {
        $bam_file->remove;
        $sorted_bam_file->remove;
        $self->sam_multi_file->remove;
    }

    return;
}

=head2 oligo_hits_three_prime_check

When looks at hits the three prime end of the oligo is critical.
So a hit on a oligo where the mismatches occur in the last 3-4 three prime bases of
the oligo should not count

This involves fetching the sequence of the alignments which can be very time consuming.

=cut
sub oligo_hits_three_prime_check {
    my ( $self, $oligo_hits ) = @_;

    my $oligo_alignments = $self->fetch_alignment_sequence( $oligo_hits );

    #TODO add 3' oligo alignment check sp12 Wed 18 Sep 2013 10:32:49 BST
    # temp code, need to replace with 3' check
    my %matches;
    for my $oligo ( keys %{ $oligo_alignments } ) {
        for my $alignment ( @{ $oligo_alignments->{ $oligo } } ) {
            if ( $alignment->{percent_hit}  == 100 ) {
                $matches{$oligo}{'exact_matches'}++;
            }
            elsif ( $alignment->{percent_hit}  >= 90 ) {
                $matches{$oligo}{'hits'}++;
            }
        }
    }

    $self->matches( \%matches );

    return;
}

=head2 fetch_alignment_sequence

Grab sequence for alignments found by bwa using fastaFromBed script.
* fastaFromBed - get sequence for each alignment

This can be a very time consuming process depending on the number of hits.

=cut
sub fetch_alignment_sequence {
    my ( $self, $oligo_hits ) = @_;

    #TODO filter out oligos with too many hits here sp12 Mon 16 Sep 2013 12:59:57 BST
    # should save a lot of time in the next step, use $oligo_hits

    #TODO output in fasta format sp12 Tue 17 Sep 2013 10:19:08 BST
    my $seq_file = $self->work_dir->file('query.seqs.tsv')->absolute;
    my @fastaFromBed_command = (
        'fastaFromBed',                       # get sequence for each alignment
        '-tab',                               # write output in tab delimited format
        '-fi', $self->target_file->stringify, # target genome file, indexed for bwa
        '-bed', $self->bed_file->stringify,   # input file of alignments
        '-fo', $seq_file->stringify,          # output file
        '-s'                                  # take notice of strand
    );
    $self->log->debug( "fastaFromBed command: " . join( ' ', @fastaFromBed_command ) );

    my ( $out, $err ) =  ( "", "" );
    run( \@fastaFromBed_command, '<', \undef, '>', \$out, '2>', \$err,)
        or DesignCreate::Exception->throw(
            "Failed to run fastaFromBed command: $err" );

    my %alignments;
    # merge bed data with seq data
    my $seq_fh = $seq_file->openr;
    my $bed_fh = $self->bed_file->openr;
    while ( my $bed_line = <$bed_fh> ) {
        my $seq_line = <$seq_fh>;

        my ( $location, $seq ) = split /\s+/, $seq_line;
        #the location needs to be further split to match the bed file format
        my ( $seq_chr, $seq_start, $seq_end ) = $location =~ /(.+):(\d+)-(\d+)/;

        my ( $chr, $start, $end, $name, $score, $strand ) = split /\s+/, $bed_line;

        #make sure the locations from each file are the same or everything would be wrong
        if ( $chr eq $seq_chr and $start == $seq_start and $end == $seq_end ) {
            push @{ $alignments{ $name } }, {
                chr         => $chr,
                start       => $start,
                end         => $end,
                strand      => $strand,
                bwa_score   => $score,
                seq         => $seq,
                percent_hit => $self->calculate_percent_alignment( $name, $seq ),
            };
        }
        else {
            die "$seq_line doesn't match $bed_line!";
        }
    }


    return \%alignments;
}

=head2 calculate_percent_alignment

Calculate the percentage hit a aligment has

=cut
sub calculate_percent_alignment {
    my ( $self, $name, $alignment_seq ) = @_;
    my $oligo_seq = $self->get_oligo_seq( $name );
    my $oligo_length = length( $oligo_seq );
    my $hamming_distance = hamming_distance( $alignment_seq, $oligo_seq );

    my $percent_hit = 100 * ( ( $oligo_length - $hamming_distance ) / $oligo_length );
    return sprintf("%.0f", $percent_hit);
}

=head2 hamming_distance

use string xor to get the number of mismatches between the two strings.
the xor returns a string with the binary digits of each char xor'd,
which will be an ascii char between 001 and 255. tr returns the number of characters replaced.

=cut
sub hamming_distance {
    die "Strings passed to hamming distance differ" if length($_[0]) != length($_[1]);
    return (uc($_[0]) ^ uc($_[1])) =~ tr/\001-\255//;
}

__PACKAGE__->meta->make_immutable;

1;

__END__
